core/headerchain.go

```diff
@@ -122,87 +122,120 @@ func (hc *HeaderChain) GetBlockNumber(hash common.Hash) *uint64 {
+		rawdb.WriteHeader(batch, header)
+
+		lastHeader = header
+		lastHash = hash
+		lastNumber = number
+		ptd = externTd
+		inserted = append(inserted, numberHash{number, hash})
+		hc.headerCache.Add(hash, header)
+		hc.numberCache.Add(hash, number)
 	}
+	batch.Write()
+	batch.Reset()
 	localTd := hc.GetTd(hc.currentHeaderHash, hc.CurrentHeader().Number.Uint64())
-	externTd := new(big.Int).Add(header.Difficulty, ptd)
-
-	// Irrelevant of the canonical status, write the td and header to the database
-	if err := hc.WriteTd(hash, number, externTd); err != nil {
-		log.Crit("Failed to write header total difficulty", "err", err)
-	}
-	rawdb.WriteHeader(hc.chainDb, header)
-
 	// If the total difficulty is higher than our known, add it to the canonical chain
 	// Second clause in the if statement reduces the vulnerability to selfish mining.
 	// Please refer to http://www.cs.cornell.edu/~ie53/publications/btcProcFC.pdf
 	if externTd.Cmp(localTd) > 0 || (externTd.Cmp(localTd) == 0 && mrand.Float64() < 0.5) {
 		// Delete any canonical number assignments above the new head
-		batch := hc.chainDb.NewBatch()
-		for i := number + 1; ; i++ {
+		for i := lastNumber + 1; ; i++ {
 			hash := rawdb.ReadCanonicalHash(hc.chainDb, i)
 			if hash == (common.Hash{}) {
 				break
 			}
 			rawdb.DeleteCanonicalHash(batch, i)
 		}
 		batch.Write()
+		batch.Reset()
 
-		// Overwrite any stale canonical number assignments
+		// Overwrite any stale canonical number assignments, work backwards
+		// from the first inserted header
 		var (
-			headHash   = header.ParentHash
-			headNumber = header.Number.Uint64() - 1
-			headHeader = hc.GetHeader(headHash, headNumber)
+			headHash   = headers[0].ParentHash
+			headNumber = headers[0].Number.Uint64() - 1
 		)
 		for rawdb.ReadCanonicalHash(hc.chainDb, headNumber) != headHash {
-			rawdb.WriteCanonicalHash(hc.chainDb, headHash, headNumber)
-
+			rawdb.WriteCanonicalHash(batch, headHash, headNumber)
+			headHeader := hc.GetHeader(headHash, headNumber)
 			headHash = headHeader.ParentHash
 			headNumber = headHeader.Number.Uint64() - 1
-			headHeader = hc.GetHeader(headHash, headNumber)
 		}
-		// Extend the canonical chain with the new header
-		rawdb.WriteCanonicalHash(hc.chainDb, hash, number)
-		rawdb.WriteHeadHeaderHash(hc.chainDb, hash)
-
-		hc.currentHeaderHash = hash
-		hc.currentHeader.Store(types.CopyHeader(header))
-
+		// Now write the current chain in a batch
+		for _, hdr := range inserted {
+			rawdb.WriteCanonicalHash(batch, hdr.hash, hdr.number)
+		}
+		rawdb.WriteHeadHeaderHash(batch, lastHash)
+		batch.Write()
+		batch.Reset()
+		hc.currentHeaderHash = lastHash
+		hc.currentHeader.Store(types.CopyHeader(lastHeader))
 		status = CanonStatTy
 	} else {
 		status = SideStatTy
 	}
-
-	hc.headerCache.Add(hash, header)
-	hc.numberCache.Add(hash, number)
-
-	return
+	// Execute any post-write callback function
+	if pwCallbackFn != nil {
+		for _, header := range headers {
```

Comment:
Can it happen that all headers are not inserted?

If so, if we should change the loop to  

```
if header.Number.Uint64() > lastNumber {
	break
}
pwCallbackFn(header, status)
```

Reply:
I guess that could happen if the imported chain is non-contiguous, with some random headers in the ending. good catch